### PR TITLE
 refactor(react): update Accordion components to hooks

### PR DESCRIPTION
**Pull request by joshblack**
Updates our Accordion components to use hooks, where applicable. Also includes general updates to tests, if you noticed a test that may be missing in the new style let me know and I can add it back in! 

#### Changelog

**New**

- `packages/react/src/prop-types/deprecate` helper for deprecating prop types

**Changed**

- `packages/react`
  - `Accordion.js` style updates to import order and class conventions
  - `Accordion-test.js` update to snapshot check
  - `Accordion.Skeleton.js` refactor logic and deprecate `uid` prop
  - `AccordionItem.js` update to use hooks
  - `AccordionItem-test.js` update to new test style and to support hooks

**Removed**

### CODE REVIEW

**Comment by jnm2377** [comment]
Question about this in general, are we actively working on refactoring components to hooks? I thought React recommended not to refactor existing class components to hooks?

**Comment by joshblack** [comment]
@jnm2377 definitely isn't an active focus, just wanted to get my feet wet with hooks during downtime 😄 

In the case of converting things, I think components that use `getDerivedStateFromProps` are probably good candidates for hook conversations. But, like you said, definitely no need to start converting right away! My hope is that we can follow their [gradual adoption strategy](https://reactjs.org/docs/hooks-intro.html#gradual-adoption-strategy)

File: packages/react/src/components/Accordion/Accordion.js [hunk 1]
```diff
@@ -5,21 +5,21 @@
...
 import PropTypes from 'prop-types';
 import React from 'react';
-import classnames from 'classnames';
-import { settings } from 'carbon-components';
 
 const { prefix } = settings;
 
-const Accordion = ({ children, className, ...other }) => {
-  const classNames = classnames(`${prefix}--accordion`, className);
+function Accordion({ children, className: customClassName, ...rest }) {
```
Comment (asudoh) [inline packages/react/src/components/Accordion/Accordion.js:15]: Any reason behind the move to function declaration here?

**Review by asudoh — COMMENTED** [review]
The code is OK, I share the thoughts with @jnm2377 on conversion. If there is imminent feature additions that would make sense. Wrt `gDSFP` great that you found the migration path, and given such straightforward migration path exists I'd be inclined pick up others.

File: packages/react/src/components/Accordion/Accordion.js (hunk 1 above)
Comment (joshblack) [reply packages/react/src/components/Accordion/Accordion.js:15]: Honestly didn't really realize it haha. I think it's from seeing components written that way when referring to the hooks docs that I subconsciously picked it up.

I do tend to like the function declaration form for components, but really the only difference in the output is a function declaration versus a function expression.

**Comment by joshblack** [comment]
@asudoh @jnm2377 in general, the biggest motivator to migrate things over to the hooks-based form would be to allow us to share more state-related helpers through custom hooks. For example, one idea was to extract out logic for using keyboard navigation to navigate a list of focusable items. 

Another, slightly more subjective, reason is that it gives us a chance to revisit older parts of the codebase that might need a refresh or update based on lessons we've learned over the years 😅

**Comment by vpicone** [comment]
I agree with @joshblack. While I'm generally against refactoring for the sake of refactoring, moving towards functional components affords us a lot of opportunities for shared shared lifecycle behavior and revisiting previous solutions for which there are more now more maintainable options. 

As an open source library, we should always be striving for the most obvious/up-to-date solutions, providing we don't sacrifice the needs of library consumers. Solutions that would allow new community members to easily grok and contribute meaningfully to components. Some of the older lifecycle methods and state management choices make that rather difficult.

**Review by emyarod — COMMENTED** [review]
looks like there is an error with the import/export changes

netlify

image

building locally

image

`Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports. Check the render method of `Container`.
Invariant Violation: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.`

**Comment by joshblack** [comment]
Ooo thanks @emyarod! Updated.

**Review by jnm2377 — APPROVED** [review]

**Review by abbeyhrt — APPROVED** [review]
Looks good to me! Thanks for starting this, it makes a great example for reference 🙂

**Review by emyarod — APPROVED** [review]
yep fixed now, looks good to me!

**Pull request merged**